UE4SS/src/LuaType/LuaTArray.cpp

```diff
@@ -223,10 +223,10 @@ namespace RC::LuaType
 
         auto array = lua_object.get_remote_cpp_object();
         auto num = array->Num();
-        if (array_index > num || num == 0)
+        if (array_index >= num || num == 0)
         {
             // We've verified with the above if-statement that we're not out-of-bounds for an int32.
```

Comment:
Is this comment misleading? `array_index` is `int64_t` and `array-Num()` returns `int32` so technically the result of `array_index - num` would be `long long`? Overflow?

Reply:
Regarding the current indexing, although not that relevant to the current PR, would it be better to
- Differentiate between a ::Get vs ::Set operation
- If it's a Get, disallow indexing outside [0, Num) and just return Lua `nil` (or nullptr UObject to be consistent with other UE4SS functions) when accessing outside the array
- If it's a Set, only allow [0, Num), plus the index after the last element for extending the array, to be set with a value, to avoid holes (if holes are required then the user must explicitly set each entry upwards, probably with `nil` or an empty). For indexes beyond that range, throw an error. Of course, this would be breaking.

Reply:
This seems reasonable.

Reply:
```lua
local tarray = p1:get()
tarray[9223372036854775807] -- max signed int64
```
```cpp
        if (array_index >= num || num == 0)
        {
            // We've verified with the above if-statement that we're not out-of-bounds for an int32.
            int32_t count = array_index == 0 || num == 0 ? 1 : static_cast<int32_t>(array_index - num + 1);
            // count = -5...
            lua_object.get_remote_cpp_object()->AddZeroed(count, lua_object.m_inner_property->GetElementSize(), Unreal::DEFAULT_ALIGNMENT);
        }
```

Additionally, when something like `tarray[9999999999999999999]` is called then cpp `int64_t test = lua.get_integer()` returns `-1`. 

My previous concerns about overflow were in regards to a ::Get() with an index greater than maxint32...

Thoughts?

Reply:
I think the bounds check I proposed above should fix this, but perhaps it should be promoted to an exception instead of a default of 0.

Reply:
Ah, didn't see the code included with your comment! Bounds check and explicit `int32_t` seem reasonable.

Reply:
I agree this should be an exception rather than default of 0. Will adjust the code accordingly.

just curious, have you considered using safeint (or something similar) for bounds checking?

Reply:
@UE4SS is there value in submitting this as a patch fix and follow with a separate breaking change for the new `:get()`/`:set()` limitations proposed-- or should I just do that as part of this PR?

I'm assuming all together makes sense unless you do something special to roll patches out faster and batch breaking changes together.

Reply:
> just curious, have you considered using safeint (or something similar) for bounds checking?

I don't know what safeint is, I've never heard of it before.



> @UE4SS is there value in submitting this as a patch fix and follow with a separate breaking change for the new `:get()`/`:set()` limitations proposed-- or should I just do that as part of this PR?
> 
> I'm assuming all together makes sense unless you do something special to roll patches out faster and batch breaking changes together.

I'm not part of the decision making group so this is just my opinion on how I think it should be done.
Please ask on the discord server if you want an official answer.

As far as I'm aware, there's no procedure in place for taking advantage of separating a breaking change and a non-breaking change into two PRs, but the people over at the discord server would know more about that since I believe internal planning and procedures are done there.

We like to keep PRs as clean as possible but at the same time we need to be reasonable with how picky we are.
I think we should accept expanding PR scopes when it makes sense as long as the PR is re-named and re-described appropriately or possibly re-created if the changes fully supersede the original PR.
If not re-created a comment should be made so that people are notified to re-read and re-review because I don't think edits send notifications.
New changes should be added to the change log when appropriate as well, something that might easily be forgotten when an existing PR is expanded instead of a new one being created.

Reply:
I think breaking changes should be kept seperate when it's not really that relevant to the existing PR. Usually we want to keep smaller breaking changes like this seperate to fixes so that fixes can go into patch release bumps and small breaking changes can go into minor release bumps. 

Since the next release will include build system change, UEPlatform and potentially Linux port, that will certainly be a minor release bump at least. While we could make an exception right here to say, "oh well it's not going to be a patch bump next release let's just bundle it in", I don't think we should set the tone on frequent exceptions just because lazy. It's not like PRs are difficult to make. I hope this makes sense! But I have indeed brought up the topic for discussion in the internal dev channel.

Reply:
@Buckminsterfullerene02 I agree with that methodology, and is my preference as well to have more smaller scoped changes; it also means you can decide when you want to merge that change independently from this bugfix :)